src/eva/vision/data/wsi/patching/coordinates.py

```diff
@@ -55,8 +57,32 @@ def from_file(
         mpp_ratio = target_mpp / level_mpp
         scaled_width, scaled_height = int(mpp_ratio * width), int(mpp_ratio * height)
 
-        for x, y in sampler.sample(scaled_width, scaled_height, wsi.level_dimensions[level_idx]):
```

Comment:
I would leave the code here unchanged, and let the sampler be in charge to only return the x,y 's that are relevant.

To achieve that, here a high level description of how I would implement this feature:

1. Add a `_mask` `@property` to `WsiDataset` which returns a numpy array that represents the mask (at the maximum wsi level (the one with lowest resolution))
2. Add optional `mask` and `min_foreground` arguments to `PatchCoordinates`, `RandomSampler` and `GridSampler` (I wouldn't write custom samplers for this, because the additional logic for doing the foreground filtering can be very minimal here, and using masks will be the default case anyways)
3. Then in the samplers, yield only the x,y coordinates of patches that exceed the `min_foreground` threshold, this would look similar to the following code for the RandomSampler as an example

```
for _ in range(self.n_samples):
    x_max, y_max = layer_shape[0], layer_shape[1]
    x, y = random.randint(0, x_max - width), random.randint(0, y_max - height)  # nosec
    patch_mask = mask[x:x + width, y:y + height]
    if mask is not None:
        if min_foreground > patch_mask.sum() / patch_mask.size:
            yield x, y
    else:
        yield x, y 
```

In `mask[x:x + width, y:y + height]` you also might need to scale the coords & width from the max level to the level used here.

What do you think? Am I missing something?